### DISCUSSION TITLE
Precheck for 'upload_to_testflight'

### DISCUSSION THREAD

**Discussion by belgiandubbel** [fastlane/fastlane#28634] [Ideas]
Feature request for **upload_to_testflight** module of fastlane to have a pre-check flag that I can run at the start of a CLI build so that I can check that it can login to AppStoreConnect successfully before starting the full build instead of it failing at the end of the build due to login issues (app specific password, 2FA token, credentials, .p8 file, etc.).

It's essentially a normal upload process, but w/o a binary file to upload.

This would make it much faster to fix CLI login issues instead of waiting for the entire build and test phase only for it to fail at the end.

**Comment by bella964** [comment]
Hello,
This is a fantastic feature idea to improve the developer workflow with Fastlane, especially for CI/CD pipelines. A pre-check flag for the upload_to_testflight module that validates login credentials and configuration before initiating the full build would save significant time and reduce frustration.

Here’s how you could frame this feature request to the Fastlane team or community:

Feature Request: Pre-check Flag for upload_to_testflight
Summary
Introduce a --pre-check (or similar) flag for the upload_to_testflight Fastlane action to validate login credentials and configurations required for connecting to App Store Connect, without requiring a binary file to upload.

Use Case
Currently, login issues with App Store Connect (e.g., invalid app-specific passwords, expired 2FA sessions, missing .p8 keys, or incorrect credentials) are only identified at the end of the build process. This can be especially problematic in CLI environments or CI/CD pipelines where full builds may take a significant amount of time, only to fail at the last step.

Adding a --pre-check flag would allow developers to:

Validate their App Store Connect login credentials at the start of the pipeline.
Quickly identify and resolve issues related to authentication or misconfiguration before proceeding with time-intensive build and test processes.
Ensure better CI/CD efficiency by reducing avoidable build failures.
Proposed Solution
Add a --pre-check flag (or equivalent parameter) to the upload_to_testflight action.
When enabled:
Perform all necessary validations for logging into App Store Connect, including:
App-specific password validity
2FA token checks (or refreshing if necessary)
Validation of .p8 file for App Store Connect API authentication
Skip the binary upload and only validate the session and credentials.
Benefits
Improved Developer Experience: Quickly identify login or configuration issues without waiting for the entire build to complete.
Time Savings: Prevents wasted time on full builds and tests when login issues exist.
Better CI/CD Efficiency: Minimizes pipeline failures caused by predictable authentication problems.
Implementation Notes
The implementation could reuse the existing login logic in upload_to_testflight but bypass any operations requiring the binary file.
Add meaningful CLI output to indicate success or specific issues with login credentials.
Optional: Expose this as a standalone action (e.g., testflight_login_check) for maximum flexibility.
Best Regards
[Maryland Health Connection com
](https://www.marylandhealth-connection.com)
